<a href="https://colab.research.google.com/github/Shourya306/EDA-Analaysis/blob/main/Copy_of_Global_Terrorism_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> The Global Terrorism Database (GTD) is an open-source database including information on terrorist attacks around the world from 1970 through 2017. The GTD includes systematic data on domestic as well as international terrorist incidents that have occurred during this time period and now includes more than 180,000 attacks. The database is maintained by researchers at the National Consortium for the Study of Terrorism and Responses to Terrorism (START), headquartered at the University of Maryland.</b>

# <b> Explore and analyze the data to discover key findings pertaining to terrorist activities. </b>

In [1]:
# Importing all the relevent libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from datetime import datetime
from datetime import date


In [2]:
pd.set_option('display.max_columns', None)     # This code is used to make pandas display all the columns in the dateset.

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
file_path = '/content/drive/MyDrive/Colab Notebooks/Module 1/M1 - Capstone project/Copy of Global Terrorism Data.csv'
# Encoded with "latin-1" because the default encoding wasn't able to encode the this large file 
df = pd.read_csv(file_path, encoding = 'latin-1', low_memory = False)        

In [8]:
df.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   

          country_txt  region                   region_txt provstate  \
0  Dominican Republic       2  Central America & Caribbean       NaN   
1              Mexico       1                North America   Federal   
2         Philippines       5               Southeast Asia    Tarlac   
3              Greece       8               Western Europe    Attica   
4               Japan       4                    East Asia   Fukouka   

            city   latitude   longitude  specificity  vicinity location  \
0  Santo Domingo  18.456792  -69.951164          1.0         0      NaN   
1    Mexico city  19.371887  -99.086624          1.0         0      NaN   
2        Unknown  15.478598  120.599741          4.0         0      NaN   
3         Athens  37.997490   23.762728          1.0         0      NaN   
4        Fukouka  33.580412  130.396361          1.0         0      NaN   

  summary  crit1  crit2  crit3  doubtterr  alternative alternative_txt  \
0     NaN      1      1      1        0.0          NaN             NaN   
1     NaN      1      1      1        0.0          NaN             NaN   
2     NaN      1      1      1        0.0          NaN             NaN   
3     NaN      1      1      1        0.0          NaN             NaN   
4     NaN      1      1      1       -9.0          NaN             NaN   

   multiple  success  suicide  attacktype1                 attacktype1_txt  \
0       0.0        1        0            1                   Assassination   
1       0.0        1        0            6     Hostage Taking (Kidnapping)   
2       0.0        1        0            1                   Assassination   
3       0.0        1        0            3               Bombing/Explosion   
4       0.0        1        0            7  Facility/Infrastructure Attack   

   attacktype2 attacktype2_txt  attacktype3 attacktype3_txt  targtype1  \
0          NaN             NaN          NaN             NaN         14   
1          NaN             NaN          NaN             NaN          7   
2          NaN             NaN          NaN             NaN         10   
3          NaN             NaN          NaN             NaN          7   
4          NaN             NaN          NaN             NaN          7   

                 targtype1_txt  targsubtype1  \
0  Private Citizens & Property          68.0   
1      Government (Diplomatic)          45.0   
2          Journalists & Media          54.0   
3      Government (Diplomatic)          46.0   
4      Government (Diplomatic)          46.0   

                                    targsubtype1_txt  \
0                                     Named Civilian   
1  Diplomatic Personnel (outside of embassy, cons...   
2                    Radio Journalist/Staff/Facility   
3                                  Embassy/Consulate   
4                                  Embassy/Consulate   

                         corp1                  target1  natlty1  \
0                          NaN             Julio Guzman     58.0   
1  Belgian Ambassador Daughter  Nadine Chaval, daughter     21.0   
2             Voice of America                 Employee    217.0   
3                          NaN             U.S. Embassy    217.0   
4                          NaN           U.S. Consulate    217.0   

          natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt  \
0  Dominican Republic        NaN           NaN           NaN              NaN   
1             Belgium        NaN           NaN           NaN              NaN   


In [9]:
df.shape

(181691, 135)

In [27]:
# Finding out the percentage of null values in an example column.
null_values_weaptype2_txt = df[df['weaptype2_txt'].isnull()]
Percentage_null_values_weaptype2_txt = (len(null_values_weaptype2_txt['weaptype2_txt']) / len(df['weaptype2_txt'])) * 100
round(Percentage_null_values_weaptype2_txt,2)

92.78

* There are **135 columns** in the dateset and it dosen't have sense to analyse all the features of the dateset because:

    1.Most of the features/columns in the dataset contain the same 
    information as the other columns (e.g weaptype1_txt & weaptype2_txt).

    2.And also these columns(not relevent columns) contain a lot of 
    null information (e.g weaptype2_txt contains 93% percent null values.


* So I will subset the above dateset with only those columns that meet the 
satisfy the above 2 conditions


In [28]:
# Subsetting the dataset.
subset_df = df[['iyear','imonth','iday','extended','country_txt','region_txt','city','latitude','longitude','success','suicide','attacktype1_txt',
                'targtype1_txt','gname','motive','weaptype1_txt','nkill','nwound']]

In [32]:
subset_df.shape

(181691, 18)

In [30]:
# Now it's time to rename the columns to make it more intuitive as to what the columns speak.
subset_df.rename(columns = {'iyear':'year','imonth':'month','iday':'day','country_txt':'country','region_txt':'region','attacktype1_txt':'attack_type',
                            'targtype1_txt':'target','gname':'organisation','weaptype1_txt':'weapon_type','nkill':'killed','nwound':'wounded'},
                 inplace = True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:5047: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [70]:
subset_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          181691 non-null  int64  
 1   day           181691 non-null  int64  
 2   extended      181691 non-null  int64  
 3   country       181691 non-null  object 
 4   region        181691 non-null  object 
 5   city          181257 non-null  object 
 6   latitude      177135 non-null  float64
 7   longitude     177134 non-null  float64
 8   success       181691 non-null  int64  
 9   suicide       181691 non-null  int64  
 10  attack_type   181691 non-null  object 
 11  target        181691 non-null  object 
 12  organisation  181691 non-null  object 
 13  motive        50561 non-null   object 
 14  weapon_type   181691 non-null  object 
 15  killed        171378 non-null  float64
 16  wounded       165380 non-null  float64
 17  month_full    181691 non-null  object 
dtypes: f

In [31]:
subset_df.head()

year  month  day  extended             country  \
0  1970      7    2         0  Dominican Republic   
1  1970      0    0         0              Mexico   
2  1970      1    0         0         Philippines   
3  1970      1    0         0              Greece   
4  1970      1    0         0               Japan   

                        region           city   latitude   longitude  success  \
0  Central America & Caribbean  Santo Domingo  18.456792  -69.951164        1   
1                North America    Mexico city  19.371887  -99.086624        1   
2               Southeast Asia        Unknown  15.478598  120.599741        1   
3               Western Europe         Athens  37.997490   23.762728        1   
4                    East Asia        Fukouka  33.580412  130.396361        1   

   suicide                     attack_type                       target  \
0        0                   Assassination  Private Citizens & Property   
1        0     Hostage Taking (Kidnapping)      Government (Diplomatic)   
2        0                   Assassination          Journalists & Media   
3        0               Bombing/Explosion      Government (Diplomatic)   
4        0  Facility/Infrastructure Attack      Government (Diplomatic)   

                         organisation motive weapon_type  killed  wounded  
0                              MANO-D    NaN     Unknown     1.0      0.0  
1  23rd of September Communist League    NaN     Unknown     0.0      0.0  
2                             Unknown    NaN     Unknown     1.0      0.0  
3                             Unknown    NaN  Explosives     NaN      NaN  
4                             Unknown    NaN  Incendiary     NaN      NaN

Before we proceed any further lets understand what each of the columns means.
  * year = The year in which attack has taken place.
  * month = The month(represented in numbers) in which the attack has taken   place.
  * day = The day in which the attack has taken place.
  * extended = If the attacked has extended for more than one day then value 1 is assigned and if it isn't then value 0 is assigned.
  * country = The country in which the incident took place.
  * city = The city in which the incident took place.
  * latitude = This feild represents the latitude.
  * longitude = This feild represents the longitude.
  * success = This feild informs us whether the attack was a success or a failure(0 = failure, 1 = Success).
  * suicide = This feild informs us whether the attack was a suicidal or not
  (0 = not suicidal, 1. = suicidal)
  * attack_type = The type of attack.
  * target = The type of target/victim.
  * organisation = Terrorist Organisation that has claimed the attack.
  * motive = The motive behind the attack.
  * weapon_type = The type of weapon used in the attack.
  * killed = The number of people killed in that attack.
  * wounded = The number of people wounded in that attack.

I have divided the **EDA** into 2 parts:
  
  * Part 1 = **Data Cleaning**
  * Part 2 = **Data Analysis**

So let's begin with **Data Cleaning**


In [52]:
# Converting the month column values into months by creating a function and then applying it on every row of the month column.
subset_df['month'].unique()

array([ 7,  0,  1,  2,  3,  4,  5,  6,  8,  9, 10, 11, 12])

In [63]:
def month_conversion(num):
  if num == 0:
     num = 'data not present'
  elif num == 1:
     num = 'jan'
  elif num == 2:
     num = 'feb'
  elif num == 3:
     num = 'mar'
  elif num == 4:
     num = 'apr'
  elif num == 5:
     num = 'may'
  elif num == 6:
     num = 'june'
  elif num == 7:
     num = 'july'
  elif num == 8:
     num = 'aug'
  elif num == 9:
     num = 'sept'
  elif num == 10:
     num = 'oct'
  elif num == 11:
     num = 'nov'
  else:
     num = 'dec'
  return num


In [66]:
subset_df['month_full'] = subset_df['month'].apply(month_conversion)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [67]:
subset_df.head()

year  month  day  extended             country  \
0  1970      7    2         0  Dominican Republic   
1  1970      0    0         0              Mexico   
2  1970      1    0         0         Philippines   
3  1970      1    0         0              Greece   
4  1970      1    0         0               Japan   

                        region           city   latitude   longitude  success  \
0  Central America & Caribbean  Santo Domingo  18.456792  -69.951164        1   
1                North America    Mexico city  19.371887  -99.086624        1   
2               Southeast Asia        Unknown  15.478598  120.599741        1   
3               Western Europe         Athens  37.997490   23.762728        1   
4                    East Asia        Fukouka  33.580412  130.396361        1   

   suicide                     attack_type                       target  \
0        0                   Assassination  Private Citizens & Property   
1        0     Hostage Taking (Kidnapping)      Government (Diplomatic)   
2        0                   Assassination          Journalists & Media   
3        0               Bombing/Explosion      Government (Diplomatic)   
4        0  Facility/Infrastructure Attack      Government (Diplomatic)   

                         organisation motive weapon_type  killed  wounded  \
0                              MANO-D    NaN     Unknown     1.0      0.0   
1  23rd of September Communist League    NaN     Unknown     0.0      0.0   
2                             Unknown    NaN     Unknown     1.0      0.0   
3                             Unknown    NaN  Explosives     NaN      NaN   
4                             Unknown    NaN  Incendiary     NaN      NaN   

         month_full  
0              july  
1  data not present  
2               jan  
3               jan  
4               jan

In [68]:
# Will be dropping the month column because it is not needed anymore.
subset_df.drop('month',axis = 1, inplace = True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [69]:
subset_df.head()

year  day  extended             country                       region  \
0  1970    2         0  Dominican Republic  Central America & Caribbean   
1  1970    0         0              Mexico                North America   
2  1970    0         0         Philippines               Southeast Asia   
3  1970    0         0              Greece               Western Europe   
4  1970    0         0               Japan                    East Asia   

            city   latitude   longitude  success  suicide  \
0  Santo Domingo  18.456792  -69.951164        1        0   
1    Mexico city  19.371887  -99.086624        1        0   
2        Unknown  15.478598  120.599741        1        0   
3         Athens  37.997490   23.762728        1        0   
4        Fukouka  33.580412  130.396361        1        0   

                      attack_type                       target  \
0                   Assassination  Private Citizens & Property   
1     Hostage Taking (Kidnapping)      Government (Diplomatic)   
2                   Assassination          Journalists & Media   
3               Bombing/Explosion      Government (Diplomatic)   
4  Facility/Infrastructure Attack      Government (Diplomatic)   

                         organisation motive weapon_type  killed  wounded  \
0                              MANO-D    NaN     Unknown     1.0      0.0   
1  23rd of September Communist League    NaN     Unknown     0.0      0.0   
2                             Unknown    NaN     Unknown     1.0      0.0   
3                             Unknown    NaN  Explosives     NaN      NaN   
4                             Unknown    NaN  Incendiary     NaN      NaN   

         month_full  
0              july  
1  data not present  
2               jan  
3               jan  
4               jan

Now lets deal with **null values**

I have replaced the motive column null values with unknown.

and also I have replaced the city column null values with unknown.

In [86]:
subset_df['motive'].value_counts()

Unknown                                                                                                                                                                                                                                                                                                                                                                  14889
The specific motive for the attack is unknown.                                                                                                                                                                                                                                                                                                                           14430
The specific motive is unknown; however, sources noted that the attack may have been part of a larger trend of violence related to Bangladesh's nationwide hartal, which began on January 6, 2015.                                                                        

In [90]:
subset_df.loc[subset_df['motive'].isnull(),'motive'].replace()

0         NaN
1         NaN
2         NaN
3         NaN
4         NaN
         ... 
181686    NaN
181687    NaN
181688    NaN
181689    NaN
181690    NaN
Name: motive, Length: 131130, dtype: object

In [91]:
subset_df['motive'].fillna('Unknown',inplace = True)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [93]:
subset_df['city'].fillna('Unknown',inplace = True)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)
